lib/ocpp/common/websocket/websocket_tls_tpm.cpp

```diff
@@ -0,0 +1,926 @@
+    size_t sent_bytes;
+    // If libwebsockets has sent all the bytes through the wire
+    volatile bool message_sent;
+};
+
+WebsocketTlsTPM::WebsocketTlsTPM(const WebsocketConnectionOptions& connection_options,
+                                 std::shared_ptr<EvseSecurity> evse_security) :
+    WebsocketBase(), evse_security(evse_security) {
+
+    set_connection_options(connection_options);
+
+    EVLOG_debug << "Initialised WebsocketTlsTPM with URI: " << this->connection_options.csms_uri.string();
+}
+
+WebsocketTlsTPM::~WebsocketTlsTPM() {
+    if (conn_data) {
+        conn_data->do_interrupt();
+    }
+
+    if (websocket_thread)
+        websocket_thread->join();
+
+    if (recv_message_thread)
+        recv_message_thread->join();
+}
+
+void WebsocketTlsTPM::set_connection_options(const WebsocketConnectionOptions& connection_options) {
+    switch (connection_options.security_profile) { // `switch` used to lint on missing enum-values
+    case security::SecurityProfile::OCPP_1_6_ONLY_UNSECURED_TRANSPORT_WITHOUT_BASIC_AUTHENTICATION:
+    case security::SecurityProfile::UNSECURED_TRANSPORT_WITH_BASIC_AUTHENTICATION:
+        throw std::invalid_argument("`security_profile` is not a TLS-profile");
+    case security::SecurityProfile::TLS_WITH_BASIC_AUTHENTICATION:
+    case security::SecurityProfile::TLS_WITH_CLIENT_SIDE_CERTIFICATES:
+        break;
+    default:
+        throw std::invalid_argument("unknown `security_profile`, value = " +
+                                    std::to_string(connection_options.security_profile));
+    }
+
+    set_connection_options_base(connection_options);
+
+    this->connection_options.csms_uri.set_secure(true);
+}
+
+static int callback_minimal(struct lws* wsi, enum lws_callback_reasons reason, void* user, void* in, size_t len) {
+    // Get user safely, since on some callbacks (void *user) can be different
+    if (nullptr != wsi) {
+        if (WebsocketTlsTPM* websocket = reinterpret_cast<WebsocketTlsTPM*>(lws_wsi_user(wsi))) {
+            return websocket->process_callback(wsi, static_cast<int>(reason), user, in, len);
+        }
+    }
+
+    return 0;
+}
+
+constexpr auto local_protocol_name = "lws-minimal-client";
+static const struct lws_protocols protocols[] = {{local_protocol_name, callback_minimal, 0, 0, 0, NULL, 0},
+                                                 LWS_PROTOCOL_LIST_TERM};
+
+static void create_sec_context(bool use_tpm, OSSL_LIB_CTX*& out_libctx, SSL_CTX*& out_ctx) {
+    OSSL_LIB_CTX* libctx = OSSL_LIB_CTX_new();
+
+    if (nullptr == libctx) {
+        EVLOG_AND_THROW(std::runtime_error("Unable to create ssl lib ctx."));
+    }
+
+    out_libctx = libctx;
+
+    if (use_tpm) {
+        OSSL_PROVIDER* prov_tpm2 = nullptr;
+        OSSL_PROVIDER* prov_default = nullptr;
+
+        if ((prov_tpm2 = OSSL_PROVIDER_load(libctx, "tpm2")) == nullptr) {
+            EVLOG_AND_THROW(std::runtime_error("Can not load provider tpm2."));
```

Comment:
```suggestion
            EVLOG_AND_THROW(std::runtime_error("Could not load provider tpm2."));
```
I suggest "could not" here, as below. Otherwise "can not" -> "cannot".